src/org/exist/storage/XQueryPool.java

```diff
@@ -58,6 +56,8 @@
 
     private final static Logger LOG = LogManager.getLogger(XQueryPool.class);
 
+    private ConcurrentMap<Source, Deque<CompiledXQuery>> pool = new ConcurrentHashMap<>();
+
     private long lastTimeOutCheck;
     private long lastTimeOfCleanup;
 
```

Comment:
`compute` methods of `ConcurrentHashMap` do not guarantee atomic access, so it'll not solve issue